drivers/storage/portworx/csi.go

```diff
@@ -44,62 +66,109 @@ func (g *csiGenerator) getSidecarContainerVersions() (csiVersions, error) {
 	k8sVer1_12, _ := version.NewVersion("1.12")
 	k8sVer1_13, _ := version.NewVersion("1.13")
 	k8sVer1_14, _ := version.NewVersion("1.14")
+	pxVer2_1, _ := version.NewVersion("2.1")
+	pxVer2_2, _ := version.NewVersion("2.2")
 
 	// Only 1.11+ is supported because only CSI 0.3 or CSI 1.x are supported
-	if g.kubeVer.LessThan(k8sVer1_11) {
+	if g.kubeVersion.LessThan(k8sVer1_11) {
 		return csiVersions{}, fmt.Errorf("kubernetes version %v does not have a "+
-			"compatible CSI version with the version of Portworx", g.kubeVer.String())
+			"compatible CSI version with the version of Portworx", g.kubeVersion.String())
 	}
 
 	var cv *csiVersions
-	// Supported only for Portworx version >= 2.1
-	// Can only use Csi 1.0 in Kubernetes 1.13+
-	if g.kubeVer.GreaterThan(k8sVer1_13) || g.kubeVer.Equal(k8sVer1_13) {
-		cv = g.setSidecarContainerVersionsV1_0()
+	if g.pxVersion.GreaterThan(pxVer2_1) || g.pxVersion.Equal(pxVer2_1) {
+		// Portworx version >= 2.1
+		// Can only use Csi 1.0 in Kubernetes 1.13+
+		if g.kubeVersion.GreaterThan(k8sVer1_13) || g.kubeVersion.Equal(k8sVer1_13) {
+			cv = g.setSidecarContainerVersionsV1_0()
+		} else {
+			cv = g.setSidecarContainerVersionsV0_4()
+		}
 	} else {
 		cv = g.setSidecarContainerVersionsV0_4()
 	}
 
+	// Check if configmaps are necessary for leader election.
+	// If it is  >=1.13.0 and and <1.14.0
+	if (g.kubeVersion.GreaterThan(k8sVer1_13) || g.kubeVersion.Equal(k8sVer1_13)) &&
+		g.kubeVersion.LessThan(k8sVer1_14) {
+		cv.includeEndpointsAndConfigMapsForLeases = true
+	}
+
+	// Feature flags for Portworx version >= 2.2
+	if g.pxVersion.GreaterThan(pxVer2_2) || g.pxVersion.Equal(pxVer2_2) {
+		// Resizing supported in portworx >= 2.2
+		cv.includeResizer = true
+	}
+
 	// Check if we need to setup the CsiNodeInfo CRD
 	// If 1.12.0 <= KubeVer < 1.14.0 create the CRD
-	if (g.kubeVer.GreaterThan(k8sVer1_12) || g.kubeVer.Equal(k8sVer1_12)) &&
-		g.kubeVer.LessThan(k8sVer1_14) {
+	if (g.kubeVersion.GreaterThan(k8sVer1_12) || g.kubeVersion.Equal(k8sVer1_12)) &&
+		g.kubeVersion.LessThan(k8sVer1_14) {
 		cv.createCsiNodeCrd = true
 	}
 
 	// Setup registration correctly
-	if g.kubeVer.LessThan(k8sVer1_13) {
+	if g.kubeVersion.LessThan(k8sVer1_13) {
 		cv.useOlderPluginsDirAsRegistration = true
 	}
 
+	// Set the CSI driver name
+	// - PX Versions <2.2.0 will always use the deprecated CSI Driver Name.
+	// - PX Versions >=2.2.0 will default to the new name
+	if g.pxVersion.LessThan(pxVer2_2) {
```

Comment:
We also need to check if the UseDeprecatedName flag is set:
https://github.com/portworx/px-installer/blob/master/px-spec-websvc/csi.go#L127

Reply:
This flag originates from the PX Installer page. If they select "Use Deprecated CSI Driver Name" (under Enable CSI), we alter the spec to use the deprecated name, as well as pass the UseDeprecatedName flag to PX. This will tell PX to use the old name.